src/passes/LLVMNontrappingFPToIntLowering.cpp

```diff
@@ -0,0 +1,185 @@
+// not semantically equivalent in wasm. This is because out-of-range inputs to
+// these instructions produce poison values. So we need only ensure that there
+// is no trap, but need not ensure any particular result. The transformation
+// in this pass is the same as the one used by LLVM to lower fptoui/fptosi
+// to wasm trapping instructions.
+
+// For example, if a conversion is guarded by a range check in the source, LLVM
+// can move the conversion before the check (and instead guard the use of the
+// result, which may be poison). This is valid in LLVM and for the nontrapping
+// wasm fptoint instructions but not for the trapping conversions. The
+// transformation in this pass is valid only if the nontrapping conversions
+// in the wasm were generated from LLVM and implement LLVM's conversion
+// semantics.
+
+namespace wasm {
+struct LLVMNonTrappingFPToIntLoweringImpl
+  : public WalkerPass<PostWalker<LLVMNonTrappingFPToIntLoweringImpl>> {
+  bool isFunctionParallel() override { return true; }
+
+  std::unique_ptr<Pass> create() override {
+    return std::make_unique<LLVMNonTrappingFPToIntLoweringImpl>();
+  }
+
+  UnaryOp getReplacementOp(UnaryOp op) {
+    switch (op) {
+      case TruncSatSFloat32ToInt32:
+        return UnaryOp::TruncSFloat32ToInt32;
+      case TruncSatUFloat32ToInt32:
+        return UnaryOp::TruncUFloat32ToInt32;
+      case TruncSatSFloat64ToInt32:
+        return UnaryOp::TruncSFloat64ToInt32;
+      case TruncSatUFloat64ToInt32:
+        return UnaryOp::TruncUFloat64ToInt32;
+      case TruncSatSFloat32ToInt64:
+        return UnaryOp::TruncSFloat32ToInt64;
+      case TruncSatUFloat32ToInt64:
+        return UnaryOp::TruncUFloat32ToInt64;
+      case TruncSatSFloat64ToInt64:
+        return UnaryOp::TruncSFloat64ToInt64;
+      case TruncSatUFloat64ToInt64:
+        return UnaryOp::TruncUFloat64ToInt64;
+      default:
+        WASM_UNREACHABLE("Unexpected opcode");
+    }
+  }
+
+  template<typename From, typename To> void replaceSigned(Unary* curr) {
+    BinaryOp ltOp;
+    UnaryOp absOp;
+    switch (curr->op) {
+      case TruncSatSFloat32ToInt32:
+      case TruncSatSFloat32ToInt64:
+        ltOp = BinaryOp::LtFloat32;
+        absOp = UnaryOp::AbsFloat32;
+        break;
+      case TruncSatSFloat64ToInt32:
+      case TruncSatSFloat64ToInt64:
+        ltOp = BinaryOp::LtFloat64;
+        absOp = UnaryOp::AbsFloat64;
+        break;
+      default:
+        WASM_UNREACHABLE("Unexpected opcode");
+    }
+
+    Builder builder(*getModule());
+    Index v = Builder::addVar(getFunction(), curr->value->type);
+    // if fabs(operand) < INT_MAX then use the trapping operation, else return
+    // INT_MIN
+    replaceCurrent(builder.makeIf(
+      builder.makeBinary(
+        ltOp,
+        builder.makeUnary(
+          absOp, builder.makeLocalTee(v, curr->value, curr->value->type)),
+        builder.makeConst(static_cast<From>(std::numeric_limits<To>::max()))),
```

Comment:
Separate question: Do you know why LLVM's code adds `-`(minus)?
https://github.com/llvm/llvm-project/blob/2906fcadb8563a02949f852867cebc63e0539cb7/llvm/lib/Target/WebAssembly/WebAssemblyISelLowering.cpp#L493

Also, does this handle when the input value is greater than INT_MIN and less than or equal to INT_MIN+1? INT_MIN's abs value would be greater than INT_MAX but it's still in the int range. (in case of int)

Reply:
Good catch, see my comment below.

Reply:
Also: I don't know why LLVM's code adds the minus. I was looking at the [tests](https://github.com/llvm/llvm-project/blob/2906fcadb8563a02949f852867cebc63e0539cb7/llvm/test/CodeGen/WebAssembly/conv-trap.ll#L12C49-L12C55) which just seem to be using 2^31 and 2^32, that's odd...

Reply:
We had a offline chat about this and I would like to leave some comments on it just for the posterity's, or very likely me reading the code later again's, sake:

- I was confused; [This LLVM code](https://github.com/llvm/llvm-project/blob/2906fcadb8563a02949f852867cebc63e0539cb7/llvm/lib/Target/WebAssembly/WebAssemblyISelLowering.cpp#L493)'s `Limit` has a minus because `INT_MIN` is a negative number.

- So the substitute numbers picked by this pass and the LLVM implementations are the same. And these values solves my question of INT_MIN handling too. The current abs comparison uses a trapping fp-to-int operations only when the abs value is less than abs(INT_MIN), which does not include INT_MIN. But in that case, the substitue value we pick is INT_MIN anyway, so the it is handled correctly.

- That being said, how about adding a one-liner comment explaining why this handles INT_MIN too?